### DISCUSSION TITLE
ChangeDetection Update Needed

### DISCUSSION THREAD

**Discussion by dskramst** [tteck/Proxmox#2522] [Help Needed]
I am not able to update my Change Detection.
I run the helper script, reboot, but it stays on the same version.

I am pretty new to this.
These helper scripts are amazing!  I have had a lot of fun learning with them!

**Comment by tteck** [comment]
I don't know of anything more to do:
https://github.com/tteck/Proxmox/blob/2171405dc6db9d683814219ac11336b8e31771c5/ct/changedetection.sh#L58-L61

**Comment by mgrimace** [comment]
@dskramst If it wasn't clear you run the word `update` from inside the LXC, not in the Proxmox shell. I literally just SSH in and type the word `update`.

@tteck is there a way to auto-run the changedetection update, similar to the option in the plex script for periodic automatic updates?

**Reply by tteck** [reply]
You can set up a cron job for _virtually_ any task.

**Reply by dskramst** [reply]
That is nifty!
I was copy/Pasting the Helper script which appears to have done the same thing, but this is easier.  Thank you.
My issue is, I "update" and it says "updated Successfully".  I open the interface (website) and it says "A new Version is Available".
It's showing V0.45.9 but on Github it shows v0.45.14 is available.
So I don't know how to get it updated

**Reply by mgrimace** [reply]
In terms of updates, mine updated to v0.45.14 using the inbuilt 'update' script - steps: SSH to change LXC, run `update`, agree. It says 'updated successful'. Maybe it's a cache issue on the webUI? (e.g., clear browser cache, restart)

**Reply by dskramst** [reply]
Oh good point.  I will try a different browser.
Is there a way from the console to see what version it's on?

**Reply by mgrimace** [reply]
`pip show changedetection.io` will show the version from the LXC console

**Comment by mgrimace** [comment]
> You can set up a cron job for _virtually_ any task.

How do you do cron with the update script since it has a dialogue? Is there an arg that we can use that would automatically agree / do it unattended?

**Reply by tteck** [reply]
Don't use the script, use the commands that actually updates the application.

**Reply by tteck** [reply]
https://github.com/tteck/Proxmox/discussions/2522#discussioncomment-8535082

**Comment by navels** [comment]
After updating you have to restart the service:
```
systemctl restart changedetection.service
```

I suppose that should added to the script.

**Reply by mgrimace** [reply]
Thanks, for automated cron updates, I now have a custom update.sh with:

```bash
#!/bin/bash

pip3 install changedetection.io 
pip3 install playwright
systemctl restart changedetection.service

```
and cron as (weekly):
`0 1 * * 0 /bin/bash /root/update.sh`

**Comment by navels** [comment]
I can't guarantee that this will work to update any LXC container . . . but it might,  it works for all of mine:
```
# get the install script
script=$(wget -qLO - "$(grep -oP 'https?://[^)]+' /usr/bin/update)")
# delete any lines after the last function definition
script=$(head -n $(grep -n '^}$' <<< "$script" | tail -n 1 | cut -d: -f1) <<< "$script")
# remove any prompts asking to update or not
script=$(awk '
  /UPD=\$\(whiptail/,/\)/ {
    if (/UPD=\$\(whiptail/) {
      print "UPD=1"
      next
    }
    next
  }
  { print }' <<< "$script")
# source what remains and run the update
source <(echo "$script")
update_script 2>/dev/null
```

**Reply by mgrimace** [reply]
Thanks for sharing this, so for a copy/paste novice, do you use this as a script that you invoke with cron?

**Reply by navels** [reply]
You could.  I prefer to know what's in the update so i subscribe the release page for the git hub repos and then just run that command when there is an update.

**Reply by stabmaster1985** [reply]
Hello @navels, can you help with your script?!

What i have to do in the first line? You add there "/usr/bin/update" ... must i replace this with the local sh skript from github?
So in first step i download the sh skript local and the replace the /usr/bin/update with my localname?

Thanks for you help.
Best regards,
Frank

**Reply by navels** [reply]
@stabmaster1985 no you just run those commands as-is.  but i want to be clear, this is only if you want to update the container non-interactively.  most of the time you would just run the install script on the container and it will prompt you to upgrade:
```
bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/changedetection.sh)"
```

**Reply by stabmaster1985** [reply]
Hey @navels ,

thanks for your reply. It worked for my use case with some code snippets. ;-)

**Comment by mgrimace** [comment]
> I am not able to update my Change Detection. I run the helper script, reboot, but it stays on the same version.
> 
> I am pretty new to this. These helper scripts are amazing! I have had a lot of fun learning with them!

I had to run `sudo apt install zlib1g-dev libjpeg-dev libpng-dev` to get the `pillow` dependency to properly update, that fixed things for me on the latest update.

**Reply by tteck** [reply]
It has been confirmed to be working. https://github.com/tteck/Proxmox/issues/2619

**Reply by navels** [reply]
FYI I don't have `libpng-dev` installed.  I do have `zlib1g-dev` but that package comes with `python3-dev` as a dependency, which is itself included in the changedetection.io install script.

**Reply by tteck** [reply]
I'm certain that some dependencies required by `python3-dev` are listed in the script, which could potentially be omitted.

**Comment by kossidts** [comment]
### Issue Encountered During Update Process
While running the `update` command, I encountered the following error:

`| Updating Change Detection LXC   
[ERROR] in line 63: exit code 0: while executing command pip3 install changedetection.io --upgrade &> /dev/null`

To investigate further, I ran the command directly in the console:

`pip3 install changedetection.io --upgrade`

This resulted in the following error:

```
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.
    
    See /usr/share/doc/python3.11/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```



Running the command with the --break-system-packages flag resolved the issue:
`pip3 install changedetection.io --upgrade --break-system-packages`

However, this approach might not be the best practice as it might break things. A potentially safer alternative would be to use pipx, which manages a virtual environment for the package and avoids interfering with the system-wide Python environment.

I hope this helps someone.

**Reply by tteck** [reply]
https://github.com/tteck/Proxmox/issues/3400
